src/public/js/rs.js

```diff
@@ -45,13 +45,214 @@ var RS = (function() {
       }
     }
     if (errors) {
-      // TODO ERROR CORRECTION
-      throw new Error('Error correction not implemented yet');
+      // Perform B-M decoding
+      var syndromes = computeSyndromePoly(receivedMsg, n, k);
+      var bm = berlekampMassey(syndromes, n, k);
+      var errorLocations = chienSearch(bm.errorLocationPoly);
+      var errorMagnitudes = forneysFormula(errorLocations,
+                                           bm.errorMagnitudePoly);
+      var repairedMsg = repair(receivedMsg, errorLocations, errorMagnitudes);
+
+      // Check that the repaired message is in fact divisible by genPoly
+      var repairedRemainder = UTIL.polynomialDiv(repairedMsg, genPoly, n, k);
+      var repairedErrors = false;
+      for (var i = 0; i < n - k; i++) {
+        if (repairedRemainder[i] != 0) {
+          throw new Error('Too many errors to decode.');
+        }
+      }
+
+      return repairedMsg.slice(0, k);
     } else {
-      return codedMsg.slice(0, k);
+      return receivedMsg.slice(0, k);
+    }
+  }
+
+  // Computes a syndrome polynomial for the specified received message
+  // by evaluating the received message at n-k consecutive powers of
+  // the generator element within Galois Field 256
+  // receivedMsg is an array of length n
+  // n specifies the length of the codeword / received message
+  // k specifies the length of the original message
+  // Returns a big-endian array of length (n-k+1) which represents a polynomial
+  // -> [1, 2, 3] is x^2 + 2*x + 3
+  function computeSyndromePoly(receivedMsg, n, k) {
+    var syndromes = new Uint8Array(n - k + 1);
+
+    // Add the 0th syndrome to represent the coefficient of z^0 even though it's not useful for error correction
+    syndromes[n - k] = 0;
+
+    var syndrome;
+    for (var i = 1; i <= n - k; i++) {
+      syndrome = UTIL.polynomialEval(receivedMsg, UTIL.exp3(i));
+      syndromes[syndromes.length - 1 - i] = syndrome;
     }
+
+    return syndromes;
   }
 
+  // Performs the Berlekamp-Massey algorithm on a syndrome polynomial for
+  // a certain n and k. Iterates (n-k) times to fit the BM "key equation".
+  // Returns an error location polynomial and an error magnitude polynomial.
+  // Input syndromePoly and returns errorLocationPoly and errorMagnitudePoly are
+  // big-endian arrays representing polynomials, i.e. [1,2,3] -> x^2 + 2x + 3
+  // Inputs n and k are integers
+  function berlekampMassey(syndromePoly, n, k) {
```

Comment:
Okay it got really late and I am not entirely solid on how this algorithm really works so I am ending the review for now. need to go over this a bit more tomorrow!